functions/go/list-setters/listsetters/list_setters_test.go

```diff
@@ -0,0 +1,394 @@
+package listsetters
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+)
+
+func TestListSettersFilter(t *testing.T) {
+	var tests = []struct {
+		name              string
+		input             string
+		kf                string
+		setterYml         string
+		expectedResources []*Result
+		errMsg            string
+		warnings          []*ErrSetterDiscovery
+	}{
+		{
+			name: "No setters",
+			input: `apiVersion: v1
+kind: Service
+metadata:
+  name: my-app
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: my-app
+  name: mungebot`,
+			expectedResources: []*Result{},
+			warnings:          []*ErrSetterDiscovery{{"unable to find Kptfile, please include --include-meta-resources flag if a Kptfile is present"}},
+		},
+		{
+			name: "Scalar Simple",
+			kf: `apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: test
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        app: my-app`,
+			input: `apiVersion: v1
+kind: Service
+metadata:
+  name: my-app # kpt-set: ${app}
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: my-app # kpt-set: ${app}
+  name: mungebot`,
+			expectedResources: []*Result{{Name: "app", Value: "my-app", Count: 2, Type: "string"}},
+		},
+		{
+			name: "Scalar Simple invalid kf",
+			kf: `apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: test
+foo: bar`,
+			input: `apiVersion: v1
+kind: Service
+metadata:
+  name: my-app # kpt-set: ${app}
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: my-app # kpt-set: ${app}
+  name: mungebot`,
+			errMsg: "unable to read Kptfile: please make sure the package has a valid 'v1' Kptfile: yaml: unmarshal errors:\n  line 8: field foo not found in type v1.KptFile",
+		},
+		{
+			name: "Scalar Simple missing apply-setters",
+			kf: `apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: test
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/foo:v0.1
+      configMap:
+        app: my-app`,
+			input: `apiVersion: v1
+kind: Service
+metadata:
+  name: my-app # kpt-set: ${app}
```

Comment:
What would be the Value in result if the current value of setters differ across fields?

Reply:
In the current implementation it would be the first discovered value (kptfile setter discovery is done first, so if there is a setter there, we use the value)

I initially had some code [here](https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/04197ada141323dd7bcecc98af5f535458b3f8d3/functions/go/list-setters/listsetters/list_setters.go#L302-L304) to emit a warning if the setter values differed.

One issue I ran into was the way `currentSetterValues` matches setters names to values. For a setter like `${platform-project-id}-${cluster-name}` with values like `platform-project-id: my-project-id`, `cluster-name: my-cluster-us-west4` individual matches become `platform-project-id: my-project-id-my-cluster-us`, `cluster-name: west4`. I couldn't find a simple way to solve this. Happy to hear if you have ideas about how we should tackle this. Example: https://play.golang.org/p/mEoPvEldYNI

FWIF I believe legacy `kpt cfg list-setters` didn't surface anything if the values differed.

Reply:
> In the current implementation it would be the first discovered value (kptfile setter discovery is done first, so if there is a setter there, we use the value)
> 

I think package consumers don't care about the current value of the setter as most of the times they are just dummy place holders like `PROJECT_ID`. So picking the first value is good enough. We can print warnings in the future based on the user feedback.

> I initially had some code [here](https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/04197ada141323dd7bcecc98af5f535458b3f8d3/functions/go/list-setters/listsetters/list_setters.go#L302-L304) to emit a warning if the setter values differed.
> 
> One issue I ran into was the way `currentSetterValues` matches setters names to values. For a setter like `${platform-project-id}-${cluster-name}` with values like `platform-project-id: my-project-id`, `cluster-name: my-cluster-us-west4` individual matches become `platform-project-id: my-project-id-my-cluster-us`, `cluster-name: west4`. I couldn't find a simple way to solve this. Happy to hear if you have ideas about how we should tackle this. Example: https://play.golang.org/p/mEoPvEldYNI
> 

Those are ambiguous cases and can't be solved. `apply-setters` throws error asking users to provide both the values if user wants to change only on of the setter values. But for list-setters case, we can have a special word like `UNRESOLVED` ? Let me know your thoughts as to what you would expect as an end user.


> FWIW I believe legacy `kpt cfg list-setters` didn't surface anything if the values differed.

Reply:
sgtm, I think first value is a good start. 

> apply-setters throws error asking users to provide both the values 

I was planning on adding a warning but when I traced through apply setters code, the way it resolved a complex value is actually like above. When I provide both `platform-project-id` and `cluster-name` it individually resolves `platform-project-id` to `my-project-id-my-cluster-us` and `cluster-name` to `west4`. The field value is still correct as the fully resolved value becomes `my-project-id-my-cluster-us-west4`.